src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala

```diff
@@ -163,6 +168,72 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
       case None =>
     }
   }
-  // scalastyle:on cyclomatic.complexity
-  // scalastyle:on method.length
+
+
+  private def updateSchema(dataFrame: DataFrame): DataFrame = {
+    var df = dataFrame
+    val ss = dataFrame.sparkSession
+    var fieldMap = Map[String, DataType]()
+
+    // If table existed before, get it's schema
+    val previousSchema: Option[StructType] = this.hudiOutputProperties.tableName match {
+      case Some(tableName) => {
+        ss.catalog.tableExists(tableName) match {
+          case true => Option(ss.table(tableName).schema)
+          case false => None
+        }
+      }
+      case None => None
+    }
+
+    // Update the schema
+    val schema = StructType(df.schema.fields.flatMap(
+      field => {
+        // Add nullability, not on by default
+        val newField = field.copy(nullable = true)
+        val fieldName = newField.name
+        var returnedFields = List[StructField]()
+
+        // By default we will remove completely null columns (will return them if existed in previous schemas), you can disable this behaviour
+        this.hudiOutputProperties.dropNullColumns match {
+          case Some(true) => {
+            // Column is detected as having only null values, we need to remove it
+            isOnlyNullColumn(df, fieldName) match {
+              case true => {
+                df = df.drop(fieldName)
+
+                // Check if removed column existed in a previous schema, if so, use the previous schema definition
+                previousSchema match {
+                  case Some(sch) => {
+                    sch.fields.filter(f => f.name == fieldName).foreach(f => fieldMap += (fieldName -> f.dataType))
+                  }
+                  case None =>
+                }
+              }
+              case false => returnedFields = returnedFields :+ newField
+            }
+          }
+          case _ => returnedFields = returnedFields :+ newField
+        }
+        returnedFields
+      })
+    )
+
+    // Update the actual schema
+    df = df.sparkSession.createDataFrame(df.rdd, schema)
+
+    // Add the new columns as null columns
+    for ((name, dataType) <- fieldMap) {
+      // scalastyle:off null
+      df = df.withColumn(name, lit(null).cast(dataType))
+      // scalastyle:on null
+    }
+    df
+  }
+
+  def isOnlyNullColumn(df: DataFrame, name: String): Boolean = {
+    df.select(name).filter(col(name).isNotNull).limit(1).count() == 0
```

Comment:
Wouldn't df.filter(col(name).isNotNull).count() == 0 be more elegant?

Reply:
It will be sufficient to know if there is 1 or 0 NotNull values, so we use the limit(1) here to get just the first filtered element, it is much faster.